src/arith/ir_mutator_with_analyzer.cc

```diff
@@ -20,13 +20,41 @@
 /*!
  * \file tvm/arith/ir_mutator_with_analyzer.cc
  */
+#include <vector>
+
 #include "ir_mutator_with_analyzer.h"
 
 #include <tvm/arith/iter_affine_map.h>
 #include <tvm/tir/analysis.h>
 #include <tvm/tir/op.h>
 
 namespace tvm {
+namespace tir {
+
+bool IsPureCondition(const PrimExpr& condition, const Stmt& scope) {
+  CallEffectKind side_effect = SideEffect(condition);
+  if (side_effect > CallEffectKind::kReadState) {
+    return false;
+  } else if (side_effect == CallEffectKind::kReadState) {
+    // when the condition maybe stateful, enter into the slow path
+    // to visit the subtree one more time and ensure the state is not mutated.
+    std::vector<const VarNode*> read_vars;
+    UsesVar(condition, [&read_vars](const VarNode* v) {
+      read_vars.push_back(v);
+      return true;
```

Comment:
By returning `true`, the `VarTouchVisitor` will return early and won't visit the rest of the TIR tree.  As a result, any read variables after the first would be missing from `read_vars`.